commit fd537bc696d65ce3baec88ecf5a825f7eaaaf484
Author: John Spray <jspray@redhat.com>
Date:   Sat Jul 16 22:04:39 2016 +0100

    mgr: enable active daemon to return to standby
    
    Rather than respawn a whole process, just fall back
    to being a standby.
    
    Signed-off-by: John Spray <john.spray@redhat.com>

diff --git a/src/mgr/MgrStandby.cc b/src/mgr/MgrStandby.cc
index 896fb166f74..0cb0cd01c1c 100644
--- a/src/mgr/MgrStandby.cc
+++ b/src/mgr/MgrStandby.cc
@@ -38,11 +38,13 @@ MgrStandby::MgrStandby() :
   active_mgr(nullptr)
 {
   client_messenger = Messenger::create_client_messenger(g_ceph_context, "mgr");
+  objecter = new Objecter(g_ceph_context, client_messenger, monc, NULL, 0, 0);
 }
 
 
 MgrStandby::~MgrStandby()
 {
+  delete objecter;
   delete monc;
   delete client_messenger;
   delete active_mgr;
@@ -82,6 +84,11 @@ int MgrStandby::init()
   client_t whoami = monc->get_global_id();
   client_messenger->set_myname(entity_name_t::CLIENT(whoami.v));
 
+  objecter->set_client_incarnation(0);
+  objecter->init();
+  client_messenger->add_dispatcher_head(objecter);
+  objecter->start();
+
   timer.init();
   send_beacon();
 
@@ -127,12 +134,40 @@ void MgrStandby::shutdown()
     active_mgr->shutdown();
   }
 
+  objecter->shutdown();
+
   timer.shutdown();
 
   monc->shutdown();
   client_messenger->shutdown();
 }
 
+void MgrStandby::handle_mgr_map(MMgrMap* mmap)
+{
+  auto map = mmap->get_map();
+  dout(4) << "received map epoch " << map.get_epoch() << dendl;
+  const bool active_in_map = map.active_gid == monc->get_global_id();
+  dout(4) << "active in map: " << active_in_map
+          << " active is " << map.active_gid << dendl;
+  if (active_in_map) {
+    if (active_mgr == nullptr) {
+      dout(1) << "Activating!" << dendl;
+      active_mgr = new Mgr(monc, client_messenger, objecter);
+      active_mgr->background_init();
+      dout(1) << "I am now active" << dendl;
+    } else {
+      dout(10) << "I was already active" << dendl;
+    }
+  } else {
+    if (active_mgr != nullptr) {
+      derr << "I was active but no longer am" << dendl;
+      active_mgr->shutdown();
+      delete active_mgr;
+      active_mgr = nullptr;
+    }
+  }
+}
+
 bool MgrStandby::ms_dispatch(Message *m)
 {
   Mutex::Locker l(lock);
@@ -140,31 +175,7 @@ bool MgrStandby::ms_dispatch(Message *m)
 
   switch (m->get_type()) {
     case MSG_MGR_MAP:
-      {
-        auto mmap = static_cast<MMgrMap*>(m);
-        auto map = mmap->get_map();
-        dout(4) << "received map epoch " << map.get_epoch() << dendl;
-        const bool active_in_map = map.active_gid == monc->get_global_id();
-        dout(4) << "active in map: " << active_in_map
-                << " active is " << map.active_gid << dendl;
-        if (active_in_map) {
-          if (active_mgr == nullptr) {
-            dout(1) << "Activating!" << dendl;
-            active_mgr = new Mgr(monc, client_messenger);
-            active_mgr->background_init();
-            dout(1) << "I am now active" << dendl;
-          } else {
-            dout(10) << "I was already active" << dendl;
-          }
-        } else {
-          if (active_mgr != nullptr) {
-            derr << "I was active but no longer am" << dendl;
-            shutdown();
-            // FIXME: should politely go back to standby (or respawn
-            // process) instead of stopping entirely.
-          }
-        }
-      }
+      handle_mgr_map(static_cast<MMgrMap*>(m));
       break;
 
     default:
@@ -174,6 +185,8 @@ bool MgrStandby::ms_dispatch(Message *m)
         return false;
       }
   }
+
+  m->put();
   return true;
 }
 
diff --git a/src/mgr/MgrStandby.h b/src/mgr/MgrStandby.h
index ccb44d74b71..4c46a0730ee 100644
--- a/src/mgr/MgrStandby.h
+++ b/src/mgr/MgrStandby.h
@@ -25,11 +25,13 @@
 #include "DaemonState.h"
 #include "ClusterState.h"
 
+class MMgrMap;
 class Mgr;
 
 class MgrStandby : public Dispatcher {
 protected:
   MonClient *monc;
+  Objecter *objecter;
   Messenger *client_messenger;
 
   Mutex lock;
@@ -39,6 +41,8 @@ protected:
 
   std::string state_str();
 
+  void handle_mgr_map(MMgrMap *m);
+
 public:
   MgrStandby();
   ~MgrStandby();
